candid_rust/src/lib.rs

```diff
@@ -0,0 +1,241 @@
+use syn::spanned::Spanned;
+
+pub mod errors;
+
+type Result<T = ()> = std::result::Result<T, errors::Error>;
+
+/// A parsing context. This is not exposed and only used internally.
+struct Context<'a> {
+    pub prog: &'a IDLProg,
+    pub declaration_map: BTreeMap<String, IDLType>,
+}
+
+impl<'a> Context<'a> {
+    pub fn new(prog: &'a IDLProg) -> Self {
+        Self {
+            prog,
+            declaration_map: BTreeMap::new(),
+        }
+    }
+
+    pub fn add_declaration(&mut self, dec: &Dec) {
+        match dec {
+            Dec::TypD(binding) => {
+                self.declaration_map
+                    .insert(binding.id.clone(), binding.typ.clone());
+            }
+            _ => {}
+        }
+    }
+
+    pub fn is_record(&self, typename: &str) -> bool {
+        if let Some(IDLType::RecordT(_)) = self.declaration_map.get(typename) {
+            true
+        } else {
+            false
+        }
+    }
+}
+
+fn idl_function_definitions(
+    context: &Context,
+    name: &str,
+    func_t: &FuncType,
+    span: &Span,
+) -> Result<TokenStream> {
+    let id = syn::Ident::new(name, span.clone());
+
+    let mut arguments_def: Vec<TokenStream> = vec![];
+
+    for (i, arg) in func_t.args.iter().enumerate() {
+        let arg_name = syn::Ident::new(&format!("arg{}", i), span.clone());
+        let ty = idl_type_usage_to_rust(context, &arg, span, false)?;
+        arguments_def.push(quote! { #arg_name: #ty });
+    }
+
+    let return_type = if func_t.rets.is_empty() {
+        quote! { () }
+    } else {
+        idl_type_usage_to_rust(
+            context,
+            &func_t.rets.first().unwrap(),
+            &Span::call_site(),
+            false,
+        )?
+    };
+
+    let return_type = if func_t.modes.contains(&FuncMode::Query) {
+        return_type
+    } else {
+        quote! { impl std::future::Future<Output = #return_type> }
+    };
+
+    Ok(quote! {
+        fn #id(#(#arguments_def),*) -> #return_type
+    })
+}
+
+fn idl_type_declaration_to_rust(
+    context: &Context,
+    ty: &Dec,
+    span: &dyn Spanned,
+) -> Result<TokenStream> {
+    let (id_str, typ) = match ty {
+        Dec::ImportD(_) => Err(errors::Error::Custom("Imports not supported.".to_string())),
+        Dec::TypD(binding) => Ok((&binding.id, &binding.typ)),
+    }?;
+
+    let id = syn::Ident::new(id_str, span.span());
+
+    // We separate variants, structs and enums from the rest, as the rest can be used
+    // as `type X = #idl_type_usage_to_rust()`
+    // but idl_type_usage_to_rust() errors on some types because they're invalid in a
+    // type usage in rust.
+    return Ok(match typ {
+        IDLType::RecordT(fields) => {
+            let mut all_fields = Vec::new();
+            for f in fields {
+                let label_str = match &f.label {
+                    Label::Id(i) => format!("id_{}", i),
+                    Label::Named(n) => n.clone(),
+                    Label::Unnamed(i) => format!("unnamed_{}", i),
```

Comment:
`Id(i)` and `Unnamed(i)` have no external difference, so the generated label should be the same, e.g. `id_{i}`.

Reply:
Done.